### ISSUE TITLE
High CPU usage when playing the radio

### ISSUE THREAD

**Issue by Solarunit** [Taiko2k/TauonMusicBox#751]
Tauon 7.2 flatpak. When playing the radio I see constant ~100% CPU load on one of the core.

**Comment by Taiko2k** [comment]
But of a hard bug since it seems to only happen randomly, but I'm pretty sure I've squashed it now.

**Comment by Solarunit** [comment]
What was the cause I am curious?

**Comment by Taiko2k** [comment]
Here's what I think was happening. There's a point in the program where's there's a loop that says, move downloaded stream data from buffer A and feed it to buffer B. But there was no minimum amount of data to wait for, so if initially there happened to be a small amount of data in A, it would start spinning in this loop moving tiny bits of data at a time. I guess because this section was written in Python, it's too inefficient when this happens, but C would have been efficient enough anyway that it wouldn't be a problem. My fix is to effectively wait for buffer A to have more data to move at a time.

**Comment by Solarunit** [comment]
>I guess because this section was written in Python, it's too inefficient when this happens

any plans to rewrite to another language?

**Comment by Taiko2k** [comment]
No, Python is perfectly capable doing this and the bug should be fixed. It's just the nature of the language, and any other languages will have their own caveats.

**Comment by Solarunit** [comment]
Still a high cpu usege imo.
HTOP when radio is playing:

![alt text](https://i.imgur.com/KRzDYGY.png)

when local file is playing:

![alt text](https://i.imgur.com/XCbCs40.png)

Could it be the same level as local files playback?

**Comment by Taiko2k** [comment]
I should be the same. I *thought* I had fixed it... can you verify what version Tauon says in about?

**Comment by Taiko2k** [comment]
Wait no, I ran into it again myself, will have to keep investigating...

**Comment by Solarunit** [comment]
7.2.1

**Comment by Taiko2k** [comment]
Okay, I gave the code a bit of a stir and committed another hopefully fix.

Rather than pushing another update, will be easier if you can manually do a partial upgrade.

Just clone this repo, then copy `t_modules/t_stream.py` to `/opt/tauon-music-box/t_modules/`, replacing the one that's there. See if that fixes it.

**Comment by Solarunit** [comment]
>Just clone this repo, then copy t_modules/t_stream.py to /opt/tauon-music-box/t_modules/, replacing the one that's there. See if that fixes it.

but I am using flatpak version

**Comment by Taiko2k** [comment]
Oh, okay you can still do it with Flatpak, just the target location is `/var/lib/flatpak/app/com.github.taiko2k.tauonmb/current/active/files/bin/t_modules/` for a Flatpak system type install, let me know if thats there for you.

**Comment by Solarunit** [comment]
Yay!

![alt text](https://i.imgur.com/FDjEpsc.png)

**Comment by Solarunit** [comment]
I still have the old one as `t_stream.py.backup`
Would it be removed automatically on the next flatpak update I am curious?

**Comment by Taiko2k** [comment]
Oh, you mean the backup one, I don't know, better remove it

**Comment by Solarunit** [comment]
done!
